test/topology_experimental_raft/test_raft_voters.py

```diff
@@ -0,0 +1,86 @@
+#
+# Copyright (C) 2024-present ScyllaDB
```

Comment:
@emdroid I'm writing in a new thread, since the discussion started in https://github.com/scylladb/scylladb/pull/21901/commits/45e49978424564f9fe8b8df680d90ec59e1a6d57#r1884326160 is now about the solution and I don't want to mess with the topic of testing.

> @patjed41 As for the number of nodes in each DC vs. the max number of voters per DC the idea was indeed that the max number of voters will se chosen so that no DC would have a majority (or a tie). Which would mean that in your example: dc1: 2 nodes, dc2: 2 nodes, dc3: 1 node and another node added to the dc1 - in this case the new node wouldn't become a voter (even though the number of nodes is lower than the voters limit).
> 
> That is in fact why I added the parametrization of [1, 3] that should also test for this use case - with the parameter of `num_nodes = 1` the DC1 would have 2 nodes and DC2 and DC3 1 node each, so the DC1 should only have 1 voter (not 2, even though the maximum allowed is higher) otherwise the test would fail (when the DC1 is shut down).

Right, I didn't notice that. I think we should have a comment in the test that explains it.

> The case with nodes removal is actually an interesting one as well and we should probably also have a test for it.

Even if we add a test with nodes removal, I'm still not sure if this is enough. There is a significant part of the solution we don't test. For example, your test with `num_nodes = 1` does indeed verify that we don't end up in the state where 2 nodes from dc1 are voters. However, it does not verify that if we added a new node to dc2 or dc3 at the end, all nodes would become voters. Also, note that the following scenarios are different:
- we add dc1 first,
- we add dc1 last or, more generally, we add 4 nodes in the test in any order where the second node from dc1 is added last.

In the first scenario, it is correct (and preferable) to make two nodes from dc1 voters initially, but we must ensure that one of the nodes stops being a voter when we add the last node. In the second scenario, it doesn't happen. We don't test the second scenario, but the first one is more complicated so maybe it's enough. However, we don't check that both nodes from dc1 where voters at the begging. We only check that we didn't lose majority at the end, but we don't check whether the voter changing logic did every step as expected.

I have a few ideas, not sure how good and not necessarily mutually exclusive:
- continue what we are doing, that is writing a few Python tests like the current one, but I think they should be stronger; for example, more topology changes and additional checking that the voter set is as expected after each topology change,
- a simple randomized test, where we do some topology changes in some dc/rack configuration and check that the voter set is as expected after each topology change,
- maybe we could abstract the voter changing logic and write unit tests in C++ (even if don't go with the distance function idea, the voter changing logic still depends only on the dc/racks properties and the current set of voters).

Reply:
OK, since you have all these things planned, I have no more complaints.

After seeing this PR, I was just worried that something is wrong here -- either I am overestimating the difficulty or you are underestimating. This PR fixes #21915, named "Add the test for the limited voters feature". It looked like this test was the only test you planned for the whole solution (now I see "integration test" in the issue description, my bad).

Anyway, I think we should be more clear about the plan in the future when we split the work on a single issue into multiple PRs. It will make reviewing easier and more effective.